### ISSUE TITLE
Having minimum size warning for control nodes in container nodes

### ISSUE THREAD

**Issue by athuyaoo** [godotengine/godot-proposals#7922]
### Describe the project you are working on

A UI-heavy RPG game

### Describe the problem or limitation you are having in your project

When developing, I noticed that the control nodes like rich text labels were unable to be seen properly. I hunted for the issue until I realized that the container node which was the parent caused it to shrink to a size of 0.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Have a warning system like missing collision shapes on collision body nodes to state how to solve the problem of invisible or zero-width control nodes in the tooltip in the event it happens

It will say to solve it:
1. Change minimum size to more than 0
2. Change the size flag in layout container sizing to expand

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

![image](https://github.com/godotengine/godot-proposals/assets/25708498/bfd7f69c-9fc9-4bf7-a3fd-1647983454f8)

it will have something like this

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This can be done with an editor plugin, but for beginners they wouldn't think to look for it, so it's better to be in the engine for UX

### Is there a reason why this should be core and not an add-on in the asset library?

This improves the user experience for those beginning to learn the Godot Engine, add ons can work, but less people will be helped by this change.

**Comment by nlupugla** [comment]
The UX for finding size 0 controls could stand to be better, but I don't think a configuration warning is the way to go. Configuration warnings are generally reserved for situations where a Node will genuinely not function as intended/expected, or needs some extra bit of configuration before it's ready to be used.

**Comment by athuyaoo** [comment]
@nlupugla 
To me that's exactly why I used the configuration warnings 😆, because a UI node which cannot be seen can be considered a malfunction and need to be changed in the editor.

If there are other ways of doing so that does not use configuration warnings, but still show the developer where to change properties clearly, I think that would be better

**Comment by athuyaoo** [comment]
@jmarceno That makes sense! What if there's a way to disable the warning so that for those who know what they're doing, they can just turn it off?

**Comment by athuyaoo** [comment]
Thank you, @jmarceno, for your insights. I appreciate your perspective on the StaticBody example, and I agree that this is a valid point to consider. Let me clarify my proposal and address some of the concerns raised:

In the Godot engine, the behavior of a RichTextLabel node and Panel node without a minimum size set does indeed work as expected in most simple cases. Users can see these UI elements without any issues. However, where my proposal comes into play is when these elements are placed within a container. In this specific scenario, the size of the UI elements can unintentionally become 0, effectively making them invisible, even though the editor visually suggests that they are fully visible due to the eye icon.

This rich text label node and panel works out of the box, 
![image](https://github.com/godotengine/godot-proposals/assets/25708498/9392cdc7-8aea-44fe-ba40-e6edbbb2a1ed)

But when I put them under the container, they suddenly disappear.
![image](https://github.com/godotengine/godot-proposals/assets/25708498/dc5cffd2-800d-4676-9f7f-512d0b3f578f)

What I think should happen is that the configuration warning will not happen in the first situation, but then when placed under a container, causing the size to become 0, the configuration warning should pop up somewhere, stating that the UI has been hidden away even though the editor shows it to be fully visible as seen by the eye icon. The warning should go away if the minimum size becomes more than 0 or the size flag is able to expand.


Now, let's address some of the points raised against my proposal:

StaticBody Explanation: It's valid to note that StaticBodies are designed to collide, and this behavior aligns with the configuration warning. However, the issue I'm highlighting is more about UI elements and UI elements are designed to be shown to the user, which means if the size is zero, this goes against the purpose of a UI element?

Workflow and Debug Issue: This may be more of a workflow/debugging issue rather than something that warrants a configuration warning. While I agree that some aspects of this could be seen as a workflow concern, it's essential to consider the user experience and potential pitfalls, especially for newcomers to the engine. If users encounter a situation where their UI becomes invisible unexpectedly, it can lead to confusion and frustration. A warning would serve as a helpful tool to guide users and prevent common pitfalls. 

Editor Configurations: Mention is made of keeping editor configurations manageable and not implying user fault. I agree that we should keep the editor as user-friendly as possible for both beginners and experienced developers.  It could be designed to appear contextually when needed and provide valuable guidance without overwhelming users, especially beginners. Furthermore, having such a warning disabled by default would ensure that it doesn't disrupt experienced users' workflows, while still being available for newcomers and those who might find it beneficial.

**Comment by athuyaoo** [comment]
> I do agree that when doing this, the expected intuitive behavior, would be for the node to show and not to disappear into the void.
Just don´t know if a warning is the path, maybe on move (add_child via editor), a node should honor the size of new child it is the only one (just thinking out loud here...don´t look to much into it)

@jmarceno that sounds great! I honestly think that would be a better idea! Basically the size wouldn't automatically turn to zero when placed in a container!

Maybe a default behaviour would be to set the minimum size automatically if within a container and the size is already set? What do you think?

**Comment by athuyaoo** [comment]
> only apply if the "real size" has become zero after the move?

Yep! That sounds like a good idea, and also only if moved within a container? A second solution could be that instead the minimum size could be set to the size of the control node before it was moved into the control.

**Comment by tivec** [comment]
> Maybe only apply if the "real size" has become zero after the move?

This makes perfect sense to me. If the flags are not set to expand, you want to be alerted *after the item moves* that your object is suddenly 0 sized. I don't know how many times this has happened to me, and this would be a QOL fix for me.

**Comment by YuriSizov** [comment]
I understand and appreciate the issue described here, but configuration warnings are not the solution here. We already have issues with a configuration warning on Control nodes that is trying to be helpful but is only annoying people because it's not always relevant and it cannot be turned off: https://github.com/godotengine/godot/issues/69895

I also don't think that we should do anything clever with the sizes when you move nodes. Your issues come from inexperience with the GUI system, but that's something that is fixed in time as you gather this experience. And once you do, anything clever that changes values under your feet is going to only get in your way. So we would be making a QoL improvement for the first comers at the cost of everyone who tries to be proficient with the engine.

**Comment by athuyaoo** [comment]
> I understand and appreciate the issue described here, but configuration warnings are not the solution here. We already have issues with a configuration warning on Control nodes that is trying to be helpful but is only annoying people because it's not always relevant and it cannot be turned off: https://github.com/godotengine/godot/issues/69895
> 
> 
> 
> I also don't think that we should do anything clever with the sizes when you move nodes. Your issues come from inexperience with the GUI system, but that's something that is fixed in time as you gather this experience. And once you do, anything clever that changes values under your feet is going to only get in your way. So we would be making a QoL improvement for the first comers at the cost of everyone who tries to be proficient with the engine.


I appreciate your perspective on this matter. You make a valid point about not wanting to make automatic adjustments that might disrupt experienced users.

Instead of altering node sizes automatically or doing the configuration warning, an alternative approach could be to leverage the existing system to provide guidance. For instance, when a control is placed within a container and only when the size has become zero, the editor could send a non-intrusive toast notification or tooltip that explains what has occurred. This way, it offers valuable guidance to beginners without impacting the workflow of proficient users. It strikes a balance by assisting newcomers while ensuring experienced users can continue working efficiently within the established system. This approach aims to enhance the overall user experience for everyone, regardless of their level of expertise with the engine.
